Use prefixes and suffixes for manifest names to indicate platform.

Hoikas opened the pull request:
This formalizes a system of prefixing and suffixing manifest names to indicate the platform. Given that the current examples of production manifests are:

`External` -> external client, windows, 32-bit
`macExternal` -> external client, macOS cider wrapper, 32-bit Intel

The baseline rule is that no prefix means Windows and no suffix means 32-bit intel. Therefore, all other operating systems and architectures must be suffixed. I included PowerPC suffixes for @dpogue whom I know is tinkering with PowerPC macs. Here is what I came up with:

Suffixes:
- ` `: (no suffix) x86
- `AMD64`: AMD64/x86_64
- `ARM`: 32-bit arm (eg NVIDIA Tegra)
- `ARM64`: 64-bit arm (eg Apple Silicon, Snapdragon X)
- `PPC`: 32-bit PowerPC (eg iMac G4)
- `PPC64`: 64-bit PowerPC (eg PowerMac G5)

Prefixes:
- ` `: (no prefix) Windows
- `macos`: macOS (**not** `mac` because this conflicts with the old cider wrapper)
- `linux`: Linux (presumably glibc based)

So, some sample manifests:
- `External64`: x64 Windows external client image manifest
- `linuxThinInternal`: x86 Linux internal client manifest
- `macosInternalPatcherARM64`: Apple Silicon macOS internal release patcher manifest

Currently, we do not build any universal/fat binaries for macOS, therefore nothing has been added for any kind of universal mac binary. I'm somewhat concerned about the UX if/when an architecture gets dropped and a user on that architecture downloads a new fat binary without that architecture. Aside from that, the best way to handle universal binaries, IMO, would be to add a `Universal` or `Fat` suffix (remember, no suffix is already established as 32-bit intel) and detect if the current binary is fat. If so, use the fat binary. If not, use the arch-specific one. This would match the Windows behavior where anyone running a 32-bit client on 64-bit Windows will continue to receive 32-bit clients, even though they can run a 64-bit client. However, as stated previously, we currently do not build universal binaries, so this is currently only a tangential point.

~~This proposal matches what I'm already doing in UruManifest - it assumes that amd64 windows binaries should go into a manifest suffixed with `64`.~~ The proposed mac changes to UruManifest will need adjustments because it wrongly reassigns the meaning of `macExternal` to "64-bit intel". There's also some other oddness in there about a mac bundle manifest that doesn't make any sense. All client files should go into the client manifest.

Hoikas committed:
Use prefixes and suffixes for manifest names to indicate platform.

Suffixes:
- ``: (no suffix) x86
- `AMD64`: AMD64/x86_64
- `ARM`: 32-bit arm (eg NVIDIA Tegra)
- `ARM64`: 64-bit arm (eg Apple Silicon, Snapdragon X)
- `PPC`: 32-bit PowerPC (eg iMac G4)
- `PPC64`: 64-bit PowerPC (eg PowerMac G5)

Prefixes:
- ``: (no prefix) Windows
- `macos`: macOS
- `linux`: Linux, presumably glibc

dpogue reviewed (commented):
Overall this looks good.

My one minor complaint (which isn't new) is that because this is hardcoded at build-time, there's no way to do something like ask plFilePatcher to download the files for another manifest to inspect them

dgelessus commented:
> Do you remember when the PhysX 2.6.4 update happened on MOULa?

Perhaps I'm misremembering - it might not have been the switch to PhysX 2.6.4, but rather when the PhysX_Setup.exe was changed from the old Ageia installer to the newer Nvidia "legacy" installer. But it definitely had something to do with the PhysX redistributable.

I have some old files from late May 2020 where I tried to retrofit the new PhysX files into the Cider wrapper (it actually worked IIRC, but I never uploaded the result anywhere). I wrote "OpenUru 159" in the file names, and Chogon's release notes for that build (from 2020-05-27) list "Physics update" as one of the changes, so that checks out.

In short: it's all at least 5 years ago.

Hoikas committed:
Allow turning off platform manifests.

Just in case someone doesn't like how I've put my finger on the scale.

Hoikas commented:
I've changed the macOS prefix to `macos` to avoid conflicting with the Cider wrapper. I've also added a configure option `USE_PLATFORM_MANIFESTS` that will turn these names off and just use the old `External`, `ThinExternal`, etc. manifest names. The new platform specific names default to on.

colincornaby commented:
> I've changed the macOS prefix to `macos` to avoid conflicting with the Cider wrapper. I've also added a configure option `USE_PLATFORM_MANIFESTS` that will turn these names off and just use the old `External`, `ThinExternal`, etc. manifest names. The new platform specific names default to on.

Does this let the macOS client connect and use the Window manifests? Not interested in that as a user feature - just would like to understand it for testing purposes.

Hoikas commented:
> Does this let the macOS client connect and use the Window manifests?

No, rather, this makes the macOS client request a different manifest that should never reference any Windows specific files.
